### ISSUE TITLE
MPEG-H support in Shaka-player

### ISSUE THREAD

**Issue by cecosme** [shaka-project/shaka-player#7656]
Dear Shaka-player team,

The Brazilian SBTVD Forum has finished the draft standards for Brazil's second-generation Digital Terrestrial Television system called TV 3.0 (more details here: https://forumsbtvd.org.br/tv3_0/#panel-phase3). The current TV 3.0 roadmap expects prototypes to be deployed in 2025/Q2-Q4 and commercial products in 2026/Q1-Q2.

The MPEG-H Audio system which is already adopted by ATSC (ATSC 3.0 TV), DVB, TTA (Korean TV) and now is adopted by Brazilian SBTVD as the sole audio codec for Brazil's TV 3.0.

I would like to open this issue to request the MPEG-H support in GStreamer. More information related to MPEG-H Audio can be found here:

Information: https://mpegh.com/news/brazil/
GitHub Decoder: https://github.com/Fraunhofer-IIS/mpeghdec
GitHub IEC Encapsulation: https://github.com/Fraunhofer-IIS/iec61937-13
Test Material: https://mpegh.com/academy/testing-and-qa/
ExoPlayer Integration: https://github.com/androidx/media/tree/main/libraries/decoder_mpegh
MediaInfo support: https://mediaarea.net/MediaInfo/ChangeLog

BR, Carlos Cosme

**Comment by joeyparrish** [comment]
> I would like to open this issue to request the MPEG-H support in GStreamer. More information related to MPEG-H Audio can be found here:

Since you filed the same ticket twice (#7655), I'm assuming you meant "Shaka Player" and not "GStreamer" above.

Codec support in a web-based player like Shaka is entirely based on the browser and the codecs it supports.  Shaka Player can't add support for a codec if a browser/device/platform doesn't have it.  Web-based players do not generally implement decoders, and rely on the browser runtime for that via HTML5 video elements and the MediaSource API.

If a browser supports MPEG-H, Shaka will play MPEG-H on that browser.

Does this help clarify how things work on the web?

**Comment by cecosme** [comment]
Thank you for clarifying this point @joeyparrish and I apologize for my mistake in the previous issue(I opened a similar issue to the GStreamer team).

Regarding the decoder part for MPEG-H, your explanation made it crystal clear, but I must highlight that MPEG-H also has an MPEG-H User Interface management part (https://github.com/Fraunhofer-IIS/mpeghdec/wiki/MPEG-H-UI-manager-example) where the viewer can select the available audio scene options authored by the content creator. Shall these MPEG-H User Interface controls be handled by Shaka Player, similar to the other controls (Captions, Fullscreen, etc) or shall this MPEG-H UI be a different layer that would overlap the HTML 5 video element and possibly other UI panels from Shaka Player? The first option seems to me to be more friendly, thinking from the viewer's quality of experience.

**Comment by joeyparrish** [comment]
> MPEG-H also has an MPEG-H User Interface management part where the viewer can select the available audio scene options authored by the content creator

Ah, I see.  Yes, those UI controls would be handled by Shaka Player.

Where would a list of audio scene options come from?  Is it described in the MPD or M3U8?

How would they be selected?  Are they in separate streams, or are they in one stream?  Is there a web API to tell the browser which scene option to decode/present from the stream?

**Comment by cecosme** [comment]
Q1. _Where would a list of audio scene options come from? Is it described in the MPD or M3U8?_
A1. This information is part of the MPEG-H bitstream (MHAS bitstream) and comes from the same container that holds the audio-compressed bitstream. Neither MPD nor M3U8 has information about it. The MPEG-H decoder detects it and raises the player's attention to handle the available options.

Q2. _How would they be selected? Are they in separate streams, or are they in one stream? Is there a web API to tell the browser which scene option to decode/present from the stream?_
A2. They are selected through the UI I mentioned in the previous post. Maybe this article (https://www.iis.fraunhofer.de/content/dam/iis/en/doc/ame/Conference-Paper/BleidtR-IEEE-2017-Development-of-MPEG-H-TV-Audio-System-for-ATSC-3-0.pdf) can clarify this question (see section named "IX. ADAPTING MPEG-H FOR CONSUMER DELIVERY").

**Comment by cecosme** [comment]
Providing additional information that may be helpful:

The UI manager (https://github.com/Fraunhofer-IIS/mpeghdec/wiki/MPEG-H-UI-manager-example) acts as a pre-processor in front of the decoder, and is in charge of extracting the Audio Scene Information and also for embedding any user interaction back into the audio stream, which is then forwarded to the MPEG-H Audio decoder.

The UI manager provides a simple two-way communication interface: 

1. To describe the Audio Scene Configuration to be used for MPEG-H user interface implementation
2. To describe the format of user settings, also called user interaction events

The XML interface is described here: https://github.com/Fraunhofer-IIS/mpeghdec/wiki/MPEG-H-UI-manager-XML-format .
Based on the implementation, if a JSON representation is preferred, a translator to/from XML is provided as well: https://github.com/Fraunhofer-IIS/mpegh-ui-translator/wiki .

**Comment by cecosme** [comment]
@absidue Q2 is accurately answered with the information I provided. Please read them again. I keep mentioning the UI manager because it is a key functional block to extracting the Audio Scene Information and receiving any user interaction (please read again my  previous post and the pointed references). 

MPEG-H is an audio codec defined in ISO/IEC 23008-3:2022, while DASH and HLS are adaptive bitrate streaming protocols, meaning DASH and HLS can carry the MPEG-H encoded bitstream. For instance, Brazil TV 3.0 is going to use DASH protocol carrying video segments using VVC, LCEVC, and audio segments using MPEG-H.

**Comment by avelad** [comment]
From what I read this is something that cannot be done right now with the current browser API. The MEPG-H standard allows this, but there is no browser that supports it yet.

**Comment by joeyparrish** [comment]
Yes, and more importantly, even when a browser has support for the codec at a basic level, this seems to require APIs between the app and the decoder that **do not exist in the web platform** today.  @cecosme, you may want to get involved in the W3C media working group to create and standardize those interfaces if you want adoption of this codec on the web.  Otherwise, I expect your codec will be limited to native apps and other platforms where direct decoder access is possible.

**Comment by cecosme** [comment]
@joeyparrish , thanks for the update!

The SBTVD Forum has specified such APIs between the app and the MPEG-H decoder for TV 3.0 Applications (DTV Play), which work in a web-environment. We are also aware of an ongoing initiative in W3C for specifying similar APIs and we fully support that activity. Here is a presentation describing the topic: https://www.w3.org/2011/webtv/wiki/images/0/0b/W3C_NGA_Interface_no_video.pdf
